### PR TITLE
Correctly inspect function calls with weird atoms

### PR DESCRIPTION
**Pull request by whatyouhide**
With this commit, we fix the inspection (both in `Macro.to_string/2` as well as `Exception.format_mfa/3`) of function calls with "weird" atoms like:

    Foo."Bar"() # was inspected as Foo.Bar()
    Foo."..."() # was inspecetd as Foo....()

and many others.

This is a rework of #5606, which was aimed at #5605.

@josevalim: I went with implementing `Inspect.Atom.classify/1`, but tailored the result to function calls specifically (`:callable`, `:non_callable`) as it turns out this is enough to reuse the logic in `Inspect.Atom.inspect/1` as well.

### CODE REVIEW

**Comment by adkron** [comment]
This looks like some great work. Is `Inspect.classify` and `Inspect.escape` intended to be part of the public API?

**Comment by whatyouhide** [comment]
@adkron currently it is not. The plan is to make them "public" inside Elixir, but not "public" as in documented and available in the public API. I don't believe we have enough use cases for this to justify the burden of adding this to the public API, but we can revisit this decision if different use cases pop up. :)

**Comment by adkron** [comment]
It feels like `classify` is pretty complicated. Should we get some tests around it if it is intended to be public?

**Comment by whatyouhide** [comment]
It is not intended to be public outside Elixir, so I would skip tests for now. We call this function in a few places, so we have a pretty good coverage anyways. Also, the logic in `classify/1` is mostly taken from `inspect/1`, which uses it now, so the tests for `inspect/1` could be enough. If others think we could benefit from tests, we can add them.

File: lib/elixir/lib/inspect.ex [hunk 1]
```diff
@@ -86,17 +84,58 @@ defimpl Inspect, for: Atom do
...
+  #
+  #   * :non_callable - an atom that cannot be used as a function call after the
+  #     . operator (for example, :<<>> is not callable because Foo.<<>> is a
+  #     syntax error); this category includes atoms like :Foo, since they are
+  #     valid identifiers but they need quotes to be used in function calls
+  #     (Foo."Bar")
+  #
+  #   * :atom - any other atom (these are usually escaped when inspected, like
+  #     :"foo and bar")
+  def classify(atom) do
```
Comment (michalmuskala) [inline lib/elixir/lib/inspect.ex]: Should this be `@doc false`?

Comment (whatyouhide) [inline lib/elixir/lib/inspect.ex]: The whole `Inspect.Atom` module is gonna be hidden as that's what happens with protocol implementations. In fact, `Inspect.Atom.inspect/1` was not part of the public API but not `@doc false`d either :)

File: lib/elixir/lib/exception.ex [hunk 2]
```diff
@@ -429,17 +429,22 @@ defmodule Exception do
...
-    case match?("\"-" <> _, fun) and String.split(fun, "-") do
-      ["\"", outer_fun, "fun", _count, "\""] ->
-        "anonymous fn#{format_arity(arity)} in #{inspect module}.#{outer_fun}"
-      _ ->
-        "#{inspect module}.#{fun}#{format_arity(arity)}"
+    fun_binary = Atom.to_string(fun)
+
+    case match?("-" <> _, fun_binary) and String.split(fun_binary, "-") do
+      ["", outer_fun, "fun", _count, ""] ->
+        "anonymous fn#{format_arity(arity)} in #{inspect(module)}.#{outer_fun}"
```
Comment (fishcakez) [inline lib/elixir/lib/exception.ex]: Do we need to classify `outer_fun` like `inspected_fun` below?

Comment (whatyouhide) [inline lib/elixir/lib/exception.ex]: I just checked and indeed we do. Good catch!

File: lib/elixir/lib/inspect.ex [hunk 3]
```diff
@@ -350,9 +389,12 @@ defimpl Inspect, for: List do
   ## Private
 
   defp key_to_binary(key) do
-    case Inspect.Atom.inspect(key) do
```
Comment (josevalim) [inline lib/elixir/lib/inspect.ex]: I believe this previous implementation was correct though? It would be slightly faster because we avoid traversing the atom multiple times?

File: lib/elixir/lib/inspect.ex [hunk 4]
```diff
@@ -86,17 +84,58 @@ defimpl Inspect, for: Atom do
...
+        :boolean_or_nil
+      valid_alias?(binary) ->
+        :alias
+      valid_atom_identifier?(binary) ->
+        if :binary.first(binary) in ?A..?Z do
+          :non_callable
+        else
+          :callable
+        end
+      atom in [:%{}, :{}, :<<>>, :..., :.., :.] ->
```
Comment (josevalim) [inline lib/elixir/lib/inspect.ex]: I believe `{}` is callable but it uses a special syntax: `Foo.{1, 2, 3}`. I don't know what we want to make out of it.

File: lib/elixir/lib/inspect.ex [hunk 5]
```diff
@@ -69,15 +69,13 @@ defimpl Inspect, for: Atom do
...
 
-    cond do
-      valid_alias?(binary) ->
+    case classify(atom) do
+      :boolean_or_nil ->
+        binary
+      :alias ->
         # If the given binary is just a succession like "Elixir.Elixir.Elixir",
         # inspect it as is, otherwise, strip the leading "Elixir".
         if only_elixir?(binary) do
```
Comment (josevalim) [inline lib/elixir/lib/inspect.ex]: I have just realized this code is also wrong. For example, `Elixir.Elixir.Atom` is being inspected as `Elixir.Atom` which is then interpreted as `Atom`. Instead of `only_elixir?(binary)`, we need to check if the first two segments is "Elixir", if so, we keep "Elixir", otherwise we remove it.

File: lib/elixir/lib/inspect.ex (hunk 4 above)
Comment (josevalim) [inline lib/elixir/lib/inspect.ex]: I am also wondering if it would make sense to make `<<>>` callable as in a tuple for making things consistent.

File: lib/elixir/lib/inspect.ex [hunk 6]
```diff
@@ -86,17 +84,58 @@ defimpl Inspect, for: Atom do
...
+        :alias
+      valid_atom_identifier?(binary) ->
+        if :binary.first(binary) in ?A..?Z do
+          :non_callable
+        else
+          :callable
+        end
+      atom in [:%{}, :{}, :<<>>, :..., :.., :.] ->
+        :non_callable
+      atom == :% or atom in Macro.binary_ops or atom in Macro.unary_ops ->
```
Comment (josevalim) [inline lib/elixir/lib/inspect.ex]: Interesting. I actually did not recall `:%` being callable. While it does include some interesting properties, such as being able to call `Atom.%(1)`, it means we can never support `Atom.%{1 => 2}` which would be consistent with `Atom.{1, 2}`. I think it is too early to allow `%` to be used as such so let's please move it to the list of non-callable above.

File: lib/elixir/lib/inspect.ex (hunk 4 above)
Comment (whatyouhide) [inline lib/elixir/lib/inspect.ex]: I think it would make sense to have `Foo.<<...>>` like `{}`. I don't know what to make out of it in regard to the changes in this PR though.

File: lib/elixir/lib/inspect.ex (hunk 3 above)
Comment (whatyouhide) [inline lib/elixir/lib/inspect.ex]: Yes it was correct. It makes sense to keep it for now.

File: lib/elixir/lib/inspect.ex (hunk 4 above)
Comment (josevalim) [inline lib/elixir/lib/inspect.ex]: For now let's continue to assume what the previous code was assuming. I.e. this line is correct for now.

**Comment by whatyouhide** [comment]
@josevalim I pushed a rework of the code in this PR, once again. Highlights:

* this inspection of functions was broken in a few places: `Macro.to_string/1`, `Exception.format_mfa/3`, `Inspect.Function`, and maybe some more that I missed
* the logic for inspecting anon funs inside other funs (`-func/arity-anonfn-count-`) was broken as well as `func` can contain `-` in it
* I added the following API: `Macro.classify_atom/1` that takes an atom or a binary and does the classification on the binary, because most of the time we pass a binary to it (as we discussed on IRC), `Inspect.Function.escape_name/1` to escape a function name, and `Inspect.Function.extract_nested_function_name_and_arity/1` to abstract the (complex-ish) logic of parsing `-func/arity-anonfn-count-`

Let me know your thoughts. The API may be bad as it is now, but at least all tests (including the failing ones I added) are green :)

File: lib/elixir/lib/inspect.ex [hunk 7]
```diff
@@ -478,12 +431,15 @@ defimpl Inspect, for: Regex do
 end
 
 defimpl Inspect, for: Function do
+  require Macro
```
Comment (josevalim) [inline lib/elixir/lib/inspect.ex]: Why is this needed?

**Comment by josevalim** [comment]
@whatyouhide it looks good to me. Ideally we would have to expose the function escape somewhere since I can see that being useful in other situations. I am not a big fan of us reaching inside protocol implementations. Although at the moment I am not quite sure where.

With atoms it is less problematic because we can always inspect them.

File: lib/elixir/lib/inspect.ex [hunk 8]
```diff
@@ -69,82 +69,35 @@ defimpl Inspect, for: Atom do
...
-          "Elixir." <> rest = binary
-          rest
+    case Macro.classify_atom(binary) do
+      :boolean_or_nil ->
+        binary
+      :alias ->
+        case binary do
+          binary when binary in ["Elixir", "Elixir.Elixir"] ->
+            binary
+          "Elixir.Elixir." <> _rest ->
```
Comment (josevalim) [inline lib/elixir/lib/inspect.ex]: :heart:

File: lib/elixir/lib/inspect.ex (hunk 7 above)
Comment (whatyouhide) [inline lib/elixir/lib/inspect.ex]: Whoopsie, leftover.

**Comment by whatyouhide** [comment]
@josevalim maybe `Inspect.Function` is not a perfect place, I agree. Maybe we can start with it and see if a different solution becomes clear in the future. Also, updated the `require Macro`.

File: lib/elixir/lib/macro.ex [hunk 9]
```diff
@@ -98,6 +98,103 @@ defmodule Macro do
...
+  binary_ops_as_strings = :lists.map(&:erlang.atom_to_binary(&1, :utf8), @binary_ops)
+
+  def classify_atom(atom) when is_atom(atom) do
+    classify_atom(Atom.to_string(atom))
+  end
+
+  def classify_atom(string) when is_binary(string) do
+    cond do
+      string in ["nil", "true", "false"] ->
+        :boolean_or_nil
```
Comment (whatyouhide) [inline lib/elixir/lib/macro.ex]: As @lexmag rightfully pointed out, this could have been split in two, `:boolean` and `:nil`, not have this `:boolean_or_nil` result. I started out with that initially actually, but then changed to `:boolean_or_nil`. The rationale for the change was: this `classify_atom/1` function is not public (so we are in control of it and don't need to be unnecessarily flexible to accommodate users needs) and for now it is only used for inspecting one way or another. When inspecting, `nil`, `true`, and `false` all belong to the same category of special atoms, that is, atoms that are not aliases but still inspected without the leading `:`.

After this whole discussion (which showed this is not the clearest design) I came up with what I think is an even better idea: we can special case `nil` and booleans only in `Inspect.Atom.inspect/1`, because they are not special when considered as function names (that is, they are just `:callable`. I will do this in a separate commit.

Comment (josevalim) [inline lib/elixir/lib/macro.ex]: :+1: for special casing it on `Inspect.Atom.inspect`. Given we are removing the distinction between boolean_or_nil, shall we rename classify_atom to classify_identifier?

Other than that, this looks good to go!

Comment (whatyouhide) [inline lib/elixir/lib/macro.ex]: Oh, great name yes. 💟

Comment (whatyouhide) [inline lib/elixir/lib/macro.ex]: Shall we rename `:atom` to `:other` or something else though then? Because it looks weird that classifying an identifier can return just "atom". Wdyt?

Comment (josevalim) [inline lib/elixir/lib/macro.ex]: We have two categories: requires escaping and does not require escaping. The ones that require escaping break into: alias, operators/tokens and others. So I think renaming :atom to :other is a good plan.

**Comment by whatyouhide** [comment]
@josevalim updated with all the mentioned changes, this should be good to go.

File: lib/elixir/lib/inspect.ex [hunk 10]
```diff
@@ -498,6 +454,32 @@ defimpl Inspect, for: Function do
...
+      :other ->
+        Inspect.Atom.escape(string_name)
+    end
+  end
+
+  # Example of this format: -func/arity-fun-count-
+  def extract_nested_function_name_and_arity("-" <> rest) do
+    ["-", count, "-", "fun", "-", arity, "/" | reversed_function] =
+      rest
+      |> :re.split("([/-])")
```
Comment (lexmag) [inline lib/elixir/lib/inspect.ex]: We can replace this call and next `Enum` call with `String.split(, ["/", "-"], trim: true)`.

Comment (whatyouhide) [inline lib/elixir/lib/inspect.ex]: Nope, we need to keep the `"/"` and `"-"` in the string. This is what the `(` and `)` do in the regex.

Comment (lexmag) [inline lib/elixir/lib/inspect.ex]: I see. 👍

**Comment by whatyouhide** [comment]
@lexmag @josevalim could you give a look at the latest commit I pushed, ab9d94f? It should work around the problem of parsing things like `-map_join/3-lists^foldl/2-0-`, which were failing otherwise (that's the failure on Travis btw).

**Comment by josevalim** [comment]
Looks good to me.
